File: robocode-tankroyale-bot-api-csharp/src/internal/BotEventHandlers.cs
using System;
using Robocode.TankRoyale.BotApi.Events;

namespace Robocode.TankRoyale.BotApi.Internal
{
  internal sealed class BotEventHandlers
  {
    readonly IBaseBot baseBot;

    // Regular bot event handlers
    internal readonly EventHandler<ConnectedEvent> onConnected = new EventHandler<ConnectedEvent>();
    internal readonly EventHandler<DisconnectedEvent> onDisconnected = new EventHandler<DisconnectedEvent>();
    internal readonly EventHandler<ConnectionErrorEvent> onConnectionError = new EventHandler<ConnectionErrorEvent>();
    internal readonly EventHandler<GameStartedEvent> onGameStarted = new EventHandler<GameStartedEvent>();
    internal readonly EventHandler<GameEndedEvent> onGameEnded = new EventHandler<GameEndedEvent>();
    internal readonly EventHandler<TickEvent> onTick = new EventHandler<TickEvent>();
    internal readonly EventHandler<SkippedTurnEvent> onSkippedTurn = new EventHandler<SkippedTurnEvent>();
    internal readonly EventHandler<DeathEvent> onDeath = new EventHandler<DeathEvent>();
    internal readonly EventHandler<DeathEvent> onBotDeath = new EventHandler<DeathEvent>();
    internal readonly EventHandler<HitBotEvent> onHitBot = new EventHandler<HitBotEvent>();
    internal readonly EventHandler<HitWallEvent> onHitWall = new EventHandler<HitWallEvent>();
    internal readonly EventHandler<BulletFiredEvent> onBulletFired = new EventHandler<BulletFiredEvent>();
    internal readonly EventHandler<BulletHitBotEvent> onHitByBullet = new EventHandler<BulletHitBotEvent>();
    internal readonly EventHandler<BulletHitBotEvent> onBulletHit = new EventHandler<BulletHitBotEvent>();
    internal readonly EventHandler<BulletHitBulletEvent> onBulletHitBullet = new EventHandler<BulletHitBulletEvent>();
    internal readonly EventHandler<BulletHitWallEvent> onBulletHitWall = new EventHandler<BulletHitWallEvent>();
    internal readonly EventHandler<ScannedBotEvent> onScannedBot = new EventHandler<ScannedBotEvent>();
    internal readonly EventHandler<WonRoundEvent> onWonRound = new EventHandler<WonRoundEvent>();
    internal readonly EventHandler<CustomEvent> onCustomEvent = new EventHandler<CustomEvent>();

    // Convenient event handlers
    internal readonly EventHandler<TickEvent> onProcessTurn = new EventHandler<TickEvent>();
    internal readonly EventHandler<TickEvent> onNewRound = new EventHandler<TickEvent>();

    // Events
    private event EventHandler<ConnectedEvent>.Subscriber OnConnected;
    private event EventHandler<DisconnectedEvent>.Subscriber OnDisconnected;
    private event EventHandler<ConnectionErrorEvent>.Subscriber OnConnectionError;
    private event EventHandler<GameStartedEvent>.Subscriber OnGameStarted;
    private event EventHandler<GameEndedEvent>.Subscriber OnGameEnded;
    private event EventHandler<TickEvent>.Subscriber OnTick;
    private event EventHandler<SkippedTurnEvent>.Subscriber OnSkippedTurn;
    private event EventHandler<DeathEvent>.Subscriber OnDeath;
    private event EventHandler<DeathEvent>.Subscriber OnBotDeath;
    private event EventHandler<HitBotEvent>.Subscriber OnHitBot;
    private event EventHandler<HitWallEvent>.Subscriber OnHitWall;
    private event EventHandler<BulletFiredEvent>.Subscriber OnBulletFired;
    private event EventHandler<BulletHitBotEvent>.Subscriber OnHitByBullet;
    private event EventHandler<BulletHitBotEvent>.Subscriber OnBulletHit;
    private event EventHandler<BulletHitBulletEvent>.Subscriber OnBulletHitBullet;
    private event EventHandler<BulletHitWallEvent>.Subscriber OnBulletHitWall;
    private event EventHandler<ScannedBotEvent>.Subscriber OnScannedBot;
    private event EventHandler<WonRoundEvent>.Subscriber OnWonRound;
    private event EventHandler<CustomEvent>.Subscriber OnCustomEvent;

    private event EventHandler<TickEvent>.Subscriber OnProcessTurn;
    private event EventHandler<TickEvent>.Subscriber OnNewRound;

    internal BotEventHandlers(IBaseBot baseBot)
    {
      this.baseBot = baseBot;

      // Regular handlers
      OnConnected += onConnected.Publish;
      OnDisconnected += onDisconnected.Publish;
      OnConnectionError += onConnectionError.Publish;
      OnGameStarted += onGameStarted.Publish;
      OnGameEnded += onGameEnded.Publish;
      OnTick += onTick.Publish;
      OnSkippedTurn += onSkippedTurn.Publish;
      OnDeath += onDeath.Publish;
      OnBotDeath += onBotDeath.Publish;
      OnHitBot += onHitBot.Publish;
      OnHitWall += onHitWall.Publish;
      OnBulletFired += onBulletFired.Publish;
      OnHitByBullet += onHitByBullet.Publish;
      OnBulletHit += onBulletHit.Publish;
      OnBulletHitBullet += onBulletHitBullet.Publish;
      OnBulletHitWall += onBulletHitWall.Publish;
      OnScannedBot += onScannedBot.Publish;
      OnWonRound += onWonRound.Publish;
      OnCustomEvent += onCustomEvent.Publish;

      // Convenient handlers
      OnProcessTurn += onProcessTurn.Publish;
      OnNewRound += onNewRound.Publish;

      // Subscribe to bot events
      onConnected.Subscribe(baseBot.OnConnected);
      onDisconnected.Subscribe(baseBot.OnDisconnected);
      onConnectionError.Subscribe(baseBot.OnConnectionError);
      onGameStarted.Subscribe(baseBot.OnGameStarted);
      onGameEnded.Subscribe(baseBot.OnGameEnded);
      onTick.Subscribe(baseBot.OnTick);
      onSkippedTurn.Subscribe(baseBot.OnSkippedTurn);
      onDeath.Subscribe(baseBot.OnDeath);
      onBotDeath.Subscribe(baseBot.OnBotDeath);
      onHitBot.Subscribe(baseBot.OnHitBot);
      onHitWall.Subscribe(baseBot.OnHitWall);
      onBulletFired.Subscribe(baseBot.OnBulletFired);
      onHitByBullet.Subscribe(baseBot.OnHitByBullet);
      onBulletHit.Subscribe(baseBot.OnBulletHit);
      onBulletHitBullet.Subscribe(baseBot.OnBulletHitBullet);
      onBulletHitWall.Subscribe(baseBot.OnBulletHitWall);
      onScannedBot.Subscribe(baseBot.OnScannedBot);
      onWonRound.Subscribe(baseBot.OnWonRound);
      onCustomEvent.Subscribe(baseBot.OnCustomEvent);
    }

    internal void FireConnectedEvent(ConnectedEvent evt)
    {
      OnConnected(evt);
    }

    internal void FireDisconnectedEvent(DisconnectedEvent evt)
    {
      OnDisconnected(evt);
    }

    internal void FireConnectionErrorEvent(ConnectionErrorEvent evt)
    {
      OnConnectionError(evt);
    }

    internal void FireGameStartedEvent(GameStartedEvent evt)
    {
      OnGameStarted(evt);
    }

    internal void FireGameEndedEvent(GameEndedEvent evt)
    {
      OnGameEnded(evt);
    }

    internal void FireSkippedTurnEvent(SkippedTurnEvent evt)
    {
      OnSkippedTurn(evt);
    }

    internal void FireTickEvent(TickEvent evt)
    {
      OnTick(evt);
    }

    internal void FireProcessTurn(TickEvent evt)
    {
      OnProcessTurn(evt);
    }

    internal void FireNewRound(TickEvent evt)
    {
      OnNewRound(evt);
    }

    internal void Fire(BotEvent evt)
    {
      switch (evt)
      {
        case TickEvent tickEvent:
          OnTick(tickEvent);
          break;

        case DeathEvent botDeathEvent:
          if (botDeathEvent.VictimId == baseBot.MyId)
            OnDeath(botDeathEvent);
          else
            OnBotDeath(botDeathEvent);
          break;

        case HitBotEvent botHitBotEvent:
          OnHitBot(botHitBotEvent);
          break;

        case HitWallEvent botHitWallEvent:
          OnHitWall(botHitWallEvent);
          break;

        case BulletFiredEvent bulletFiredEvent:
          OnBulletFired(bulletFiredEvent);
          break;

        case BulletHitBotEvent bulletHitBotEvent:
          if (bulletHitBotEvent.VictimId == baseBot.MyId)
            OnHitByBullet(bulletHitBotEvent);
          else
            OnBulletHit(bulletHitBotEvent);
          break;

        case BulletHitBulletEvent bulletHitBulletEvent:
          OnBulletHitBullet(bulletHitBulletEvent);
          break;

        case BulletHitWallEvent bulletHitWallEvent:
          OnBulletHitWall(bulletHitWallEvent);
          break;

        case ScannedBotEvent scannedBotEvent:
          OnScannedBot(scannedBotEvent);
          break;

        case SkippedTurnEvent skippedTurnEvent:
          OnSkippedTurn(skippedTurnEvent);
          break;

        case WonRoundEvent wonRoundEvent:
          OnWonRound(wonRoundEvent);
          break;

        case CustomEvent customEvent:
          OnCustomEvent(customEvent);
          break;

        default:
          Console.Error.WriteLine("Unhandled event: " + evt);
          break;
      }
    }
  }
}

File: robocode-tankroyale-bot-api-java/src/main/java/dev/robocode/tankroyale/botapi/internal/BotEventHandlers.java
package dev.robocode.tankroyale.botapi.internal;

import dev.robocode.tankroyale.botapi.IBaseBot;
import dev.robocode.tankroyale.botapi.events.*;

final class BotEventHandlers {

  final IBaseBot baseBot;

  // Regular bot event handlers
  final EventHandler<ConnectedEvent> onConnected = new EventHandler<>();
  final EventHandler<DisconnectedEvent> onDisconnected = new EventHandler<>();
  final EventHandler<ConnectionErrorEvent> onConnectionError = new EventHandler<>();
  final EventHandler<GameStartedEvent> onGameStarted = new EventHandler<>();
  final EventHandler<GameEndedEvent> onGameEnded = new EventHandler<>();
  final EventHandler<RoundStartedEvent> onRoundStarted = new EventHandler<>();
  final EventHandler<RoundEndedEvent> onRoundEnded = new EventHandler<>();
  final EventHandler<TickEvent> onTick = new EventHandler<>();
  final EventHandler<SkippedTurnEvent> onSkippedTurn = new EventHandler<>();
  final EventHandler<DeathEvent> onDeath = new EventHandler<>();
  final EventHandler<DeathEvent> onBotDeath = new EventHandler<>();
  final EventHandler<HitBotEvent> onHitBot = new EventHandler<>();
  final EventHandler<HitWallEvent> onHitWall = new EventHandler<>();
  final EventHandler<BulletFiredEvent> onBulletFired = new EventHandler<>();
  final EventHandler<BulletHitBotEvent> onHitByBullet = new EventHandler<>();
  final EventHandler<BulletHitBotEvent> onBulletHit = new EventHandler<>();
  final EventHandler<BulletHitBulletEvent> onBulletHitBullet = new EventHandler<>();
  final EventHandler<BulletHitWallEvent> onBulletHitWall = new EventHandler<>();
  final EventHandler<ScannedBotEvent> onScannedBot = new EventHandler<>();
  final EventHandler<WonRoundEvent> onWonRound = new EventHandler<>();
  final EventHandler<CustomEvent> onCustomEvent = new EventHandler<>();

  final EventHandler<TickEvent> onNextTurn = new EventHandler<>();

  BotEventHandlers(IBaseBot baseBot) {
    this.baseBot = baseBot;

    onConnected.subscribe(baseBot::onConnected);
    onDisconnected.subscribe(baseBot::onDisconnected);
    onConnectionError.subscribe(baseBot::onConnectionError);
    onGameStarted.subscribe(baseBot::onGameStarted);
    onGameEnded.subscribe(baseBot::onGameEnded);
//    onRoundStarted.subscribe(baseBot::onRoundStarted); // TODO
//    onRoundEnded.subscribe(baseBot::onRoundEnded); // TODO
    onTick.subscribe(baseBot::onTick);
    onSkippedTurn.subscribe(baseBot::onSkippedTurn);
    onDeath.subscribe(baseBot::onDeath);
    onBotDeath.subscribe(baseBot::onBotDeath);
    onHitBot.subscribe(baseBot::onHitBot);
    onHitWall.subscribe(baseBot::onHitWall);
    onBulletFired.subscribe(baseBot::onBulletFired);
    onHitByBullet.subscribe(baseBot::onHitByBullet);
    onBulletHit.subscribe(baseBot::onBulletHit);
    onBulletHitBullet.subscribe(baseBot::onBulletHitBullet);
    onBulletHitWall.subscribe(baseBot::onBulletHitWall);
    onScannedBot.subscribe(baseBot::onScannedBot);
    onWonRound.subscribe(baseBot::onWonRound);
    onCustomEvent.subscribe(baseBot::onCustomEvent);
  }

  void fire(BotEvent event) {
    if (event instanceof TickEvent) {
      onTick.publish((TickEvent) event);
    } else if (event instanceof ScannedBotEvent) {
      onScannedBot.publish((ScannedBotEvent) event);
    } else if (event instanceof SkippedTurnEvent) {
      onSkippedTurn.publish((SkippedTurnEvent) event);
    } else if (event instanceof HitBotEvent) {
      onHitBot.publish((HitBotEvent) event);
    } else if (event instanceof HitWallEvent) {
      onHitWall.publish((HitWallEvent) event);
    } else if (event instanceof BulletFiredEvent) {
      onBulletFired.publish((BulletFiredEvent) event);
    } else if (event instanceof BulletHitWallEvent) {
      onBulletHitWall.publish((BulletHitWallEvent) event);
    } else if (event instanceof BulletHitBotEvent) {
      BulletHitBotEvent bulletEvent = (BulletHitBotEvent) event;
      if (bulletEvent.getVictimId() == baseBot.getMyId()) {
        onHitByBullet.publish((BulletHitBotEvent) event);
      } else {
        onBulletHit.publish((BulletHitBotEvent) event);
      }
    } else if (event instanceof DeathEvent) {
      DeathEvent deathEvent = (DeathEvent) event;
      if (deathEvent.getVictimId() == baseBot.getMyId()) {
        onDeath.publish((DeathEvent) event);
      } else {
        onBotDeath.publish((DeathEvent) event);
      }
    } else if (event instanceof BulletHitBulletEvent) {
      onBulletHitBullet.publish((BulletHitBulletEvent) event);
    } else if (event instanceof RoundStartedEvent) {
      onRoundStarted.publish((RoundStartedEvent) event);
    } else if (event instanceof RoundEndedEvent) {
      onRoundEnded.publish((RoundEndedEvent) event);
    } else if (event instanceof WonRoundEvent) {
      onWonRound.publish((WonRoundEvent) event);
    } else if (event instanceof CustomEvent) {
      onCustomEvent.publish((CustomEvent) event);
    } else {
      throw new IllegalStateException("Unhandled event type: " + event);
    }
  }
}